### PR TITLE
Team search: add sorting

### PR DESCRIPTION
**Pull request by smatting**
Follow-up for #1286. This PR adds ordering of the search result by name, email, handle, samlIdp, managedBy to `/teams/:tid/search`.

This PR requires rebuilding the ES index. However this PR doesn't increase the migration version, since this PR is meant to be released together with #1286.

### CODE REVIEW

File: libs/wire-api/src/Wire/API/Team/Role.hs [hunk 1]
```diff
@@ -106,6 +108,21 @@ instance FromJSON Role where
     -- ~fisx, Wed Jan 23 16:38:52 CET 2019
     bad -> fail $ "not a role: " <> show bad
 
+instance ToByteString Role where
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Team/Role.hs:111]: pls add round-trip test.

Comment (fisx) [reply libs/wire-api/src/Wire/API/Team/Role.hs:111]: (and pls make sure the arbitrary instance doesn't need to be changed if new constructors are added (something involving `[minBound..]`))

Comment (fisx) [reply libs/wire-api/src/Wire/API/Team/Role.hs:111]: also, i would implement the aeson instances in terms of the bytestring instances, now that you have the latter.  but feel free to disagree.

File: libs/wire-api/src/Wire/API/User/Search.hs [hunk 2]
```diff
@@ -199,3 +205,58 @@ instance FromJSON TeamContact where
...
+  = SortByName
+  | SortByHandle
+  | SortByEmail
+  | SortBySAMLIdp
+  | SortByManagedBy
+  | SortByRole
+  | SortByCreatedAt
+  deriving (Show, Eq, Ord)
+
+instance ToByteString TeamUserSearchSortBy where
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/User/Search.hs:219]: round-trip test.  (there are more, but i'll let you find those.  :)

File: services/brig/src/Brig/User/Search/Index.hs [hunk 3]
```diff
@@ -440,7 +440,10 @@ indexMapping =
                   mpIndex = True,
                   mpAnalyzer = Nothing,
                   mpFields =
-                    Map.fromList [("prefix", MappingField MPText "prefix_index" "prefix_search")]
+                    Map.fromList
+                      [ ("prefix", MappingField MPText (Just "prefix_index") (Just "prefix_search")),
+                        ("keyword", MappingField MPKeyword Nothing Nothing)
+                      ]
```
Comment (fisx) [inline services/brig/src/Brig/User/Search/Index.hs:446]: can you talk me through this on the phone?

File: services/brig/src/Brig/User/Search/TeamUserSearch.hs [hunk 4]
```diff
@@ -66,25 +68,27 @@ teamUserSearch tid mbSearchText _mRoleFilter _mSortBy _mSortOrder (fromRange ->
               searchResults = results
             }
 
--- FUTUREWORK: Maybe (sortby=<name|handle|email|saml_idp|managed_by|role|created_at>, Maybe ES.SortOrder)
--- FUTUREWORK: Maybe [Role]
--- analogous to SearchIndex.hs
+-- FUTURWORK: Implement role filter (needs galley data)
```
Comment (fisx) [inline services/brig/src/Brig/User/Search/TeamUserSearch.hs:71]:
```suggestion
 -- analogous to SearchIndex.hs
-- FUTURWORK: Implement role filter (needs galley data)
```

(unless you removed that on purpose?  if so then i'm ok with it being gone.)

File: services/brig/test/integration/API/TeamUserSearch.hs [hunk 5]
```diff
@@ -72,11 +76,48 @@ testEmptyQuerySorted :: TestConstraints m => Brig -> m ()
...
     assertEqual
       "user ids"
       (sort (fmap userId users <> [ownerId]))
       (sort (fmap teamContactUserId r))
   liftIO $ assertEqual "sorted team contacts" (sortOn Down creationDates) creationDates
+
+testSort :: TestConstraints m => Brig -> m ()
+testSort brig = do
+  (tid, userId -> ownerId, usersImplicitOrder) <- createPopulatedBindingTeamWithNamesAndHandles brig 4
+  users <- liftIO $ shuffleM usersImplicitOrder
```
Comment (fisx) [inline services/brig/test/integration/API/TeamUserSearch.hs:91]: `shuffleM` has no effect: you're never using `users` in the order you establish here, you always run `sort` on it first.  (this means you can also change `package.yaml` back, I think.)

File: services/brig/test/integration/API/TeamUserSearch.hs [hunk 6]
```diff
@@ -72,11 +76,48 @@ testEmptyQuerySorted :: TestConstraints m => Brig -> m ()
...
+              userId
+              ( case sortOrder of
+                  SortOrderAsc -> sortOn orderProp users
+                  SortOrderDesc -> sortOn (Down . orderProp) users
+              )
+      r <- searchResults <$> executeTeamUserSearch brig tid ownerId Nothing Nothing (Just tuSortBy) (Just sortOrder)
+      let rUids = filter (/= ownerId) $ fmap teamContactUserId r
+      liftIO $ assertEqual ("users sorted by " <> cs (toByteString tuSortBy)) uids rUids
+
+-- Creating test users for these cases seems overkill just to test sorting.
```
Comment (fisx) [inline services/brig/test/integration/API/TeamUserSearch.hs:114]: i don't understand this line.

**Review by fisx — CHANGES_REQUESTED** [review]
I think this is all good, just minor nit-picks and me not taking the time to read ES docs.  Let's have a phone call!

File: services/brig/test/integration/API/TeamUserSearch.hs (hunk 5 above)
Comment (smatting) [reply services/brig/test/integration/API/TeamUserSearch.hs:91]: There seems to be an implicit ordering for 1) users in cassandra 2) users in Elasticsearch. The shuffle exist to mitigate this effect. Without this the test succeeds even if you don't sort in some cases.

Comment (smatting) [reply services/brig/test/integration/API/TeamUserSearch.hs:91]: - [ ] comment this

File: services/brig/test/integration/API/TeamUserSearch.hs (hunk 6 above)
Comment (smatting) [reply services/brig/test/integration/API/TeamUserSearch.hs:114]: - [ ] reforumulate

File: libs/wire-api/src/Wire/API/Team/Role.hs (hunk 1 above)
Comment (smatting) [reply libs/wire-api/src/Wire/API/Team/Role.hs:111]: Done, with separate tests for both: aeson and bytestring

File: libs/wire-api/src/Wire/API/User/Search.hs (hunk 2 above)
Comment (smatting) [reply libs/wire-api/src/Wire/API/User/Search.hs:219]: Added bytestring rountrip tests for all new types:
- Role
- TeamUserSearchSortOrder
- TeamUserSearchSortBy
- RoleFilter

Added aeson roundtrip tests for all new types:
- TeamContact
- SearchResult TeamContact

**Comment by smatting** [comment]
I've dropped "collaborator" as serialization for `RoleExternalPartner` @fisx is this okay?

File: libs/wire-api/src/Wire/API/Team/Role.hs [hunk 7]
```diff
@@ -108,12 +108,10 @@ instance ToByteString Role where
...
-    takeLazyByteString >>= \case
-      "owner" -> pure RoleOwner
-      "admin" -> pure RoleAdmin
-      "member" -> pure RoleMember
-      "partner" -> pure RoleExternalPartner
-      bad -> fail ("not a role:  " <> show bad)
+    RoleOwner <$ string "owner"
+      <|> RoleAdmin <$ string "admin"
+      <|> RoleMember <$ string "member"
+      <|> RoleExternalPartner <$ string "partner"
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Team/Role.hs:114]: I wonder about the error message now?

**Review by fisx — APPROVED** [review]

**Pull request merged**